File: backend/app/schemas/wine.py
from typing import Annotated, Optional

from pydantic import BaseModel, Field, field_validator

VALID_TYPES = {"red", "white", "rosé", "sparkling", "dessert", "fortified"}
RatingLevel = Annotated[int, Field(ge=1, le=5)]


class WineUpdate(BaseModel):
    name: Optional[str] = None
    winery: Optional[str] = None
    vintage: Optional[int] = None
    type: Optional[str] = None
    country: Optional[str] = None
    region: Optional[str] = None
    subregion: Optional[str] = None
    abv: Optional[Annotated[float, Field(ge=0, le=100)]] = None
    sweetness: Optional[RatingLevel] = None
    acidity: Optional[RatingLevel] = None
    tannin: Optional[RatingLevel] = None
    body: Optional[RatingLevel] = None
    fruitiness: Optional[RatingLevel] = None
    description: Optional[str] = None
    image_url: Optional[str] = None

    @field_validator("type")
    @classmethod
    def _validate_type(cls, value):
        if value is None:
            return value
        normalized = value.lower()
        if normalized not in VALID_TYPES:
            raise ValueError(f"invalid type {value!r}")
        return normalized


class GrapeOut(BaseModel):
    name: str
    percentage: Optional[float] = None


class WineListItem(BaseModel):
    id: int
    name: str
    winery: str
    vintage: Optional[int] = None
    type: str
    country: Optional[str] = None
    region: Optional[str] = None
    grapes: list[GrapeOut]
    price: Optional[float] = None
    currency: Optional[str] = None
    price_usd_approx: Optional[float] = None
    image_url: Optional[str] = None
    sweetness: Optional[int] = None
    acidity: Optional[int] = None
    tannin: Optional[int] = None
    body: Optional[int] = None
    fruitiness: Optional[int] = None


class WineListResponse(BaseModel):
    total: int
    items: list[WineListItem]


class RetailerListingOut(BaseModel):
    retailer: str
    price: Optional[float] = None
    currency: Optional[str] = None
    price_usd_approx: Optional[float] = None
    product_url: Optional[str] = None
    availability: Optional[str] = None


class WineDetail(WineListItem):
    subregion: Optional[str] = None
    abv: Optional[float] = None
    description: Optional[str] = None
    listings: list[RetailerListingOut]

File: frontend/src/types/wine.ts
export interface Grape {
  name: string;
  percentage: number | null;
}

export interface WineListItem {
  id: number;
  name: string;
  winery: string;
  vintage: number | null;
  type: string;
  country: string | null;
  region: string | null;
  grapes: Grape[];
  price: number | null;
  currency: string | null;
  price_usd_approx: number | null;
  image_url: string | null;
  sweetness: number | null;
  acidity: number | null;
  tannin: number | null;
  body: number | null;
  fruitiness: number | null;
}

export interface WineListResponse {
  total: number;
  items: WineListItem[];
}

export interface RetailerListing {
  id: number;
  retailer: string;
  price: number | null;
  currency: string | null;
  price_usd_approx: number | null;
  product_url: string | null;
  availability: string | null;
}

export interface WineDetail extends WineListItem {
  subregion: string | null;
  abv: number | null;
  description: string | null;
  listings: RetailerListing[];
}

export type SortOption = "price_asc" | "price_desc" | "vintage" | "winery";

export interface ListWinesParams {
  q?: string;
  type?: string;
  country?: string;
  grape?: string;
  min_price?: number;
  max_price?: number;
  sort?: SortOption;
  limit?: number;
  offset?: number;
}

export interface RecommendationAnswers {
  sweetness?: number | number[];
  acidity?: number | number[];
  tannin?: number | number[];
  body?: number | number[];
  fruitiness?: number | number[];
  type?: string;
  country?: string;
  min_price?: number;
  max_price?: number;
}

export interface RecommendationRequest extends RecommendationAnswers {
  limit?: number;
  offset?: number;
}

export interface RecommendationItem extends WineListItem {
  match_score: number;
  explanation: string[];
}

export interface RecommendationResponse {
  profile: { description: string[] };
  total: number;
  items: RecommendationItem[];
}

export interface ColumnStats {
  count: number;
  null_count: number;
  min: number | null;
  max: number | null;
  avg: number | null;
}

export interface AdminStats {
  wines_count: number;
  wineries_count: number;
  retailers_count: number;
  listings_count: number;
  numeric: Record<string, ColumnStats>;
  categorical: Record<string, Record<string, number>>;
}

export interface WineUpdatePayload {
  name?: string;
  winery?: string;
  vintage?: number | null;
  type?: string;
  country?: string | null;
  region?: string | null;
  subregion?: string | null;
  abv?: number | null;
  sweetness?: number | null;
  acidity?: number | null;
  tannin?: number | null;
  body?: number | null;
  fruitiness?: number | null;
  description?: string | null;
  image_url?: string | null;
}

export interface ListingCreatePayload {
  retailer: string;
  price?: number | null;
  currency?: string | null;
  availability?: string | null;
  product_url?: string | null;
}

export interface ListingUpdatePayload {
  price?: number | null;
  currency?: string | null;
  availability?: string | null;
  product_url?: string | null;
}

export interface GrapeInput {
  name: string;
  percentage: number | null;
}
